Unity-Technologies/ScriptableRenderPipeline
# Keywords

### Purpose of this PR
Adds user defined Keywords to the graph via the Blackboard. These can be used to wrap graph generated code in ifdefs for performance and compilation reasons.

#### Goals
- Must provide a full featured user keyword system
- Must support predefined keywords such as HDRP’s material quality
- Must be explicit, code readability is a secondary concern
- Must provide minimal interpolators per keyword permutation
- Must allow keyword definition inside Sub Graphs

#### Design
- Create and Edit Keywords on the Blackboard
- Have sections of the Blackboard for Properties and Keywords
- Keywords come in two types (Boolean and Enum), this cannot be changed after creation
- Predefined Keyword definition allows you to switch in the graph based on Keywords you know exist in the pipeline
- Exposing Keywords allows you to set them from the Material inspector
- Dragging Keywords onto the Graph creates Keyword nodes
- Keyword nodes cannot be changed to reference other Keywords (like properties)
- Keyword nodes have one input port per Keyword option
- Keyword nodes update automatically to reflect changes on the Keyword (in the Blackboard)

#### Implementation notes:
- Each node is wrapped in all active permutations individually
   - This is because evaluating the graph per branch in a way that generates code similar to how a human would write the branches is a LOT of work, this method was deemed OK as it was explicit and added less code than writing the full graph per permutation
   - Later we will improve this by evaluating permutations that share lots of code and attempting to patch them together
- Interpolators in HDRP
   - Interpolator structs are if/def in the same manner as the main graph code. This is the ideal for now
   - Packing structs are much more complicated because of how they are made. Therefore for now they written in full, once per permutation.
- Interpolators in URP
   - These are currently not handled. This is because we have a ES customer bug to fix an issue that requires moving URP to use active fields for interpolators (like HDRP). Supporting the current interpolator generation in URP was deemed a waste of time, we will just do the active fields work soon.
- Sub Graphs
   - Keywords inside Sub Graphs are treated the same as regular Keywords.
   - To expose a Keyword from inside a Sub Graph to the Material Inspector you must duplicate it into the main graph and expose it.

#### QA Notes:
- All property refactor work landed in another PR so risk to Properties should be minimal, but still give them a check, especially copying/reordering both Keywords and Properties on the Blackboard.
- Make sure to really test every edge case of deleting/reordering Enum Keyword entries while their nodes on the graph have edges connected. This was prone to failure during development.
- Make sure to check both Generated and Preview code of the graph while testing. Correct scoping of ifdefs in the code is central to the success of this PR.
- This feature has a complicated integration with Sub Graphs. Make sure to test these two features together vigorously.

Full spec: https://favro.com/organization/c564ede4ed3337f7b17986b6/d06b8de3e40da6d340b5b063?card=Uni-86690

---
### Release Notes
See Changelog

---
### Testing status
**Katana Tests**: Coming...

**Manual Tests**: What did you do?
- Other: 

**Automated Tests**: What did you setup? (Add a screenshot or the reference image of the test please)
Coming...

Any test projects to go with this to help reviewers?
Coming...

---
### Overall Product Risks
**Technical Risk**: None, Low, Medium, High? High - Large feature with lots of moving parts. Large affect on shader generation.

**Halo Effect**: None, Low, Medium, High? Medium - Large risk to shader generation. Most of the risk to properties and blackboard landed in an earlier PR.

---
### Comments to reviewers
See Favro card linked above.
